### PR TITLE
[dagster-wandb] Integration with Weights & Biases

### PR DESCRIPTION
**Pull request by chrishiste**
Co-authored-by: Dave La Chasse <906387+dlachasse@users.noreply.github.com>
Co-authored-by: Ashraf Shaik <105249796+moonshot-s@users.noreply.github.com>

Note: W&B partners with the Dagster team on this. If you are eager to use this integration for your own projects, please hold while we get this polished. Don't hesitate to provide feedback or ask questions :) 

### Summary & Motivation
This PR contains the integration code with [Weights & Biases](https://docs.wandb.ai/).

Making it easy within Dagster to:
- use and create W&B [Artifacts](https://docs.wandb.ai/guides/artifacts)
- use and create Registered Models in W&B [Model Registry](https://docs.wandb.ai/guides/models)
- run training jobs on dedicated compute using W&B [Launch](https://docs.wandb.ai/guides/launch)
- use the [wandb](https://github.com/wandb/wandb) client in ops and assets

You will find code for the integration and examples on how to use the integration.

For more in depth documentation check the [current draft](https://docs.google.com/document/d/10qOyhbJKnJR4kazGqMF22db6UhXpeHop3_QV90eaiW4/edit).
 
**Useful references:** 
- https://docs.wandb.ai/ref/python/artifact
- https://docs.wandb.ai/ref/python/run
- https://docs.wandb.ai/guides/launch

**Notes and questions for the Dagster team:**
- We're using the local storage to store downloaded Artifacts. Does Dagster always run with a local filesystem that we can use? And what would happen when there is no more disk space?
- We linked to `https://docs.dagster.io/integrations/wandb` in the `examples/with_wandb/README.md`. Can you confirm it's where we documentation would eventually live?
- I'm expecting your review to be mostly on the form over the content but we're open to refactor the code to suit Dagster coding style better. Don't hesitate to be vocal if something could be improved.

### How I Tested These Changes
This integration has been extensively tested manually and through unit tests (included in the code).

**Known Issues**
tox fails currently due to a protobuf bug. There is an open Github [issue](https://github.com/protocolbuffers/protobuf/issues/10075). I was able to remove the error by pinning another version but I'm not sure of the implications. Probably better to wait for a fix until we release this.

### CODE REVIEW

File: python_modules/dagster/setup.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def get_version() -> str:
             "grpcio-tools>=1.32.0,<1.44.0",  # related to above grpcio pins
             "mock==3.0.5",
             "objgraph",
-            "protobuf==3.13.0",  # without this, pip will install the most up-to-date protobuf
+            "protobuf==3.20.3",  # without this, pip will install the most up-to-date protobuf
```
Comment (chrishiste) [inline python_modules/dagster/setup.py:102]: Note: this is updating the dagster module itself.

File: python_modules/dagster-wandb/dagster_wandb/io_manager.py [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+class ArtifactsIOManager(IOManager):
+    """IO Manager to handle Artifacts in Weights & Biases (W&B) .
+
+    It handles 3 different inputs:
+    - Pickable objects (the serialization module is configurable)
+    - W&B Objects (Audio, Table, Image, etc)
+    - W&B Artifacts
+    """
+
+    def __init__(self, wandb_client, config: Config):
```
Comment (yuhan) [inline python_modules/dagster-wandb/dagster_wandb/io_manager.py:91]: we could add a `base_dir` arg here to allow users to specify their own `base_dir`; if it's not provided by the user, we can default to the dagster system base dir. see in [1]

File: python_modules/dagster-wandb/dagster_wandb/io_manager.py [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+        "dagster_run_id": dagster_run_id,
+        "wandb_host": wandb_host,
+        "wandb_entity": wandb_entity,
+        "wandb_project": wandb_project,
+        "wandb_run_name": wandb_run_name,
+        "wandb_run_id": wandb_run_id,
+        "wandb_run_tags": wandb_run_tags,
+        "cache_duration_in_minutes": cache_duration_in_minutes,
+    }
+    return ArtifactsIOManager(wandb_client, config)
```
Comment (yuhan) [inline python_modules/dagster-wandb/dagster_wandb/io_manager.py:565]: [1] so base_dir can be threaded in here like:
```
base_dir = init_context.resource_config.get(
        "base_dir", init_context.instance.storage_directory() if init_context.instance else None
    )
```

File: python_modules/dagster-wandb/dagster_wandb/io_manager.py [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+
+    wandb_run_name = None
+    wandb_run_id = None
+    wandb_run_tags = None
+    cache_duration_in_minutes = None
+    if context.resource_config is not None:
+        wandb_run_name = context.resource_config.get("run_name")
+        wandb_run_id = context.resource_config.get("run_id")
+        wandb_run_tags = context.resource_config.get("run_tags")
+        cache_duration_in_minutes = context.resource_config.get("cache_duration_in_minutes")
```
Comment (yuhan) [inline python_modules/dagster-wandb/dagster_wandb/io_manager.py:548]: let's put these in `config_schema` to io manager for schematization
```
@io_manager(
  ...
  config_schema=...
)
```
example: https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-duckdb/dagster_duckdb/io_manager.py#L79-L83

**Review by yuhan — COMMENTED** [review]
Thanks a ton for putting this together. Overall it looks great!

A few feedback around the example:
1. could we add a tests dir `examples/with_wandb/with_wandb_tests` in the example so we can include the unit tests all together in our CI. example: https://github.com/dagster-io/dagster/blob/master/examples/development_to_production/development_to_production_tests
2. naming wise, i tend to use nouns for asset names (which usually represent some objects like artifacts, tables, models, etc) and verbs for op names (which often are operations therefore verbs)

i saw the tests are failing likely related to our set up for new module or example. i'll take a stab and try fixing those in this branch.

**Review by chrishiste — COMMENTED** [review]
@yuhan Thanks for the review. I've pushed another commit including your feedback. Please have a look 😄

File: python_modules/dagster-wandb/dagster_wandb/resources.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            StringSource,
+            description="The Weights & Biases host server to connect to",
+            is_required=False,
+            default_value=WANDB_CLOUD_HOST,
+        ),
+    },
+    description="Resource for interacting with Weights & Biases",
+)
+def wandb_resource(context: InitResourceContext) -> Dict[str, Any]:
+    """Resource for interacting with Weights & Biases."""
```
[the remark this answers is not in the document]
Comment (yuhan) [reply python_modules/dagster-wandb/dagster_wandb/resources.py]: the nice thing about including the code snippets here is it'd show up in the api docs directly, for example: https://docs.dagster.io/_apidocs/libraries/dagster-airbyte so the reader doesn't need to navigate away

File: examples/with_wandb/with_wandb/assets/advanced_example.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    ],
+}
+
+MY_ASSET = "my_advanced_artifact"
+MY_TABLE = "my_table"
+
+
+@asset(
+    name=MY_ASSET,
+    compute_kind="Python",
```
[the remark this answers is not in the document]
Comment (yuhan) [reply examples/with_wandb/with_wandb/assets/advanced_example.py:33]: I think W&B would look better in asset graphs and more clear that it's an integration asset (imagining a graph with tags like W&B, dbt, airbyte, etc all together so it provides a bird's eye view of the entire data platform)

[the remark this answers is not in the document]
Comment (chrishiste) [reply examples/with_wandb/with_wandb/assets/advanced_example.py]: Alright, fair enough ;)

File: examples/with_wandb/with_wandb/assets/example/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+This directory is used to showcase how to include files and folders in a W&B Artifact created by the integration.
```
[the remark this answers is not in the document]
Comment (chrishiste) [reply examples/with_wandb/with_wandb/assets/example/README.md:1]: @yuhan I think you missed this comment :)

File: python_modules/dagster-wandb/dagster_wandb/resources.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrishiste) [reply python_modules/dagster-wandb/dagster_wandb/resources.py]: I see that makes sense I will create a commit including this, will let you know when it's ready

**Review by yuhan — COMMENTED** [review]
> We're using the local storage to store downloaded Artifacts. Does Dagster always run with a local filesystem that we can use? And what would happen when there is no more disk space?

dagster is always running with a local filesystem, but when it's deployed in a remote setup, for example, each run could be in its own container which means local filesystem will be ephemera in a run scope, i.e. files downloaded to run 1's local filesystem won't be available to the next run.

i think it's ok to default to local filesystem, but it'd be great to give an option for users to download to other filesystems, e.g. an option to hook to an s3 bucket. -- but this isn't blocking this PR IMO. we can add it in a later iteration.

File: examples/with_wandb/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# Using Dagster with Weights & Biases Examples
+
+This directory contains examples showcasing the integration with Weights & Biases (W&B).
+
+View this example in the Dagster docs at https://docs.dagster.io/integrations/wandb.
```
[the remark this answers is not in the document]
Comment (chrishiste) [reply examples/with_wandb/README.md:5]: Good call, I changed all those links to point to the full docs on our side

File: python_modules/dagster/setup.py (hunk 1 above)
Comment (chrishiste) [reply python_modules/dagster/setup.py:112]: Without this change, tests won't run due to incompatible versions. I can move this out to another PR. Should I do this? 
Then we merge that single line change. Before this one. 
@yuhan

Comment (yuhan) [reply python_modules/dagster/setup.py:112]: ah i see. lets include it then.

**Review by yuhan — APPROVED** [review]
Thank you again for putting this together and bearing with my slow reviews. I can take it from here and aim to include in the next release : )

**Comment by chrishiste** [comment]
@yuhan Glad to hear it, thanks for the help :)

File: python_modules/dagster/setup.py (hunk 1 above)
Comment (yuhan) [reply python_modules/dagster/setup.py:112]: ok finally removed the pin here in a separate PR: https://github.com/dagster-io/dagster/pull/11974

**Comment by yuhan** [comment]
cherry-picked commits from https://github.com/dagster-io/dagster/pull/11830 here to fix the tests + merge latest master into this branch to prep for merging :)

**Comment by yuhan** [comment]
the follow up is https://github.com/dagster-io/dagster/pull/12059 which converts the example to using the latest Definitions apis

**Pull request merged**